### PR TITLE
Adding first example

### PR DESCRIPTION
**Pull request by chrislovecnm**
This PR includes the source code for the tutorial that is hosted here: https://bazel-contrib.github.io/SIG-rules-authors/go-tutorial.html

The code for the above HTML is on the gh-pages branch of the SIG-rules-authors repository:
https://github.com/bazel-contrib/SIG-rules-authors/tree/gh-pages

This tutorial source code includes:

- basic go rules support
- a basic cobra CLI application
- unit test
- Gazelle support

After this PR is finalized, I will update the tutorial markdown.

We are also adding CI/CD for the example.  This PR modifies the
presubmit.yml file.  This project also uses the words file that
requires the installation of "wamerican", and we are installing that
package during CI/CD.

Fixes: https://github.com/bazelbuild/rules_go/issues/3316

### CODE REVIEW

**Comment by chrislovecnm** [comment]
So we are having issues with the CLA, which we have, but I use an alias email and not my login email (security). Secondly, I don't understand why CI is running in the examples directory 🤔  - It is running in the wrong directory.

```
(16:51:43) ERROR: Analysis of target '//examples/go-code-tutorial:go-code-tutorial' failed; build aborted: no such package 'cmd': BUILD file not found in any of the following directories. Add a BUILD file to a directory to mark it as a package.
--
  | - /workdir/cmd
```

@fmeum @alexeagle

**Comment by fmeum** [comment]
> So we are having issues with the CLA, which we have, but I use an alias email and not my login email (security).

Could you sign the CLA again, this time with your alias? That's going to be the easiest solution by far as we don't control the CLA checks and would otherwise need to get Googlers involved.

> Secondly, I don't understand why CI is running in the examples directory thinking - It is running in the wrong directory.

Adding `/examples` to the top-level `.bazelignore` should help.

**Comment by chrislovecnm** [comment]
We have a corporate CLA, and my email is set to my alias - we may need to get a googler involved. I may try another PR with the GitHub hidden email 🤔

**Comment by fmeum** [comment]
Yes, please try changing your committer email first. If that doesn't work, I can ping someone.

**Comment by alexeagle** [comment]
when you add a nested workspace, you need to exclude that folder from the outer workspace, Bazel does not do that automatically. You should add `examples` to `.bazelignore`

(note, it also excludes it to add a `local_repository` pointing the outer workspace to the inner one, but IMO that introduces what appears like a load-bearing edge in the graph where it's actually dead code)

**Comment by chrislovecnm** [comment]
@fmeum I have the correct email address matched to the commit.  So I am not certain why the Google CLA machine is not happy

**Comment by chrislovecnm** [comment]
@fmeum so the CI passed, but of course, it did not test the code.  How do we get CI to do a bazel call in another dir?

**Comment by fmeum** [comment]
@chrislovecnm Your committer email is `c...e@cnmconsulting.net`. Are you sure you signed the CLA for that using your current GitHub account and have that email visible? If not, does this help:

> You may have [Keep my email address private](https://github.com/settings/emails#:~:text=Keep%20my%20email%20addresses%20private) enabled. Without a visible email address, the CLA cannot be checked. Uncheck it and re-create the offending commit, or have your CLA point of contact add your @users.noreply.github.com address to the CLA group.

You can add a new task to `presubmit.yml` and use the `working_directory` setting (see https://github.com/bazelbuild/rules_go/blob/d517cd9bec43f0541b421315fe633cdd0f204270/.bazelci/presubmit.yml#L30 for an example).

**Comment by chrislovecnm** [comment]
@fmeum awesome I will do that.  I do need someone from Google to check on the CLA.  I have set it up six ways sideways :(

**Comment by alexeagle** [comment]
Squashing to a single commit might make it easier, since all of them need to pass the validation

File: .bazelci/presubmit.yml
Comment (chrislovecnm) [inline .bazelci/presubmit.yml:40]: I'm gonna add an issue to have the macOS build work for this, but I have zero ideas on how to configure it, and the buildkite site was not helpful.

Comment (chrislovecnm) [reply .bazelci/presubmit.yml:40]: https://github.com/bazelbuild/rules_go/issues/3318 is filed

**Comment by chrislovecnm** [comment]
@fmeum and @alexeagle PTAL!

File: examples/go-code-tutorial/README.md
Comment (fmeum) [inline examples/go-code-tutorial/README.md:6]: How about moving this part to a separate file called `CONTRIBUTING.md` or `DEVELOPERS.md`? It's meant for rules_go developers, not for the typical audience of this tutorial.

File: examples/go-code-tutorial/WORKSPACE
Comment (fmeum) [inline examples/go-code-tutorial/WORKSPACE:29]: This comment should explain that "here" means something along the lines of "in `deps.bzl%go_dependencies`, which is called here so that it wins over Gazelle" - otherwise it's a bit surprising not to see any actual `go_repository`s below.

File: examples/go-code-tutorial/cmd/roll.go
Comment (fmeum) [inline examples/go-code-tutorial/cmd/roll.go:1]: Please set a consistent license header everywhere (see https://github.com/bazelbuild/rules_go/blob/d517cd9bec43f0541b421315fe633cdd0f204270/go/tools/bazel/runfiles.go#L1). https://github.com/google/addlicense makes this easy.

**Comment by fmeum** [comment]
Looks pretty good to me, just left a few comments. I would also like to have @alexeagle take a look before merging.

**Comment by chrislovecnm** [comment]
@alexeagle PTAL

File: .bazelci/presubmit.yml
Comment (alexeagle) [inline .bazelci/presubmit.yml:46]: this seems really surprising, can you add a comment here about why we have to install custom packages on the machine outside of Bazel? it makes it harder for users to reproduce the same result locally, doesn't it?

File: examples/go-code-tutorial/.gitignore
Comment (alexeagle) [inline examples/go-code-tutorial/.gitignore:1]: this file is redundant with the root .gitignore, is there a reason for that?

File: examples/go-code-tutorial/BUILD.bazel
Comment (alexeagle) [inline examples/go-code-tutorial/BUILD.bazel:1]: nit: it seems a bit surprising to me that "tutorial" is in the name of this folder, I'd expect a lot of users will just look in rules_go and be happy to find an example, but not know or care that there's also a tutorial accompanying it.

File: examples/go-code-tutorial/cmd/roll.go
Comment (alexeagle) [inline examples/go-code-tutorial/cmd/roll.go:45]: users aren't here to learn Cobra, why bother with all this text?

File: examples/go-code-tutorial/deps.bzl
Comment (alexeagle) [inline examples/go-code-tutorial/deps.bzl:1]: seems to me that all these files need commenting so the example can stand on its own to illustrate how things work.

and/or link to the tutorial from all those comments so the two are mutually reinforcing

File: .bazelci/presubmit.yml
Comment (chrislovecnm) [reply .bazelci/presubmit.yml:46]: You only have to install it if you are on LInux.  I am happy to take this as a bug and rewrite this.  We would need to change the source code and also the tutorial.

File: examples/go-code-tutorial/.gitignore
Comment (chrislovecnm) [reply examples/go-code-tutorial/.gitignore:1]: it is part of the tutorial.  I talk about setting up the ignore file

File: examples/go-code-tutorial/BUILD.bazel
Comment (chrislovecnm) [reply examples/go-code-tutorial/BUILD.bazel:1]: I suck at naming 😆 - Do you have a better one?

File: examples/go-code-tutorial/deps.bzl
Comment (chrislovecnm) [reply examples/go-code-tutorial/deps.bzl:1]: Yah, I noticed this.  Let me update the file more.

File: .bazelci/presubmit.yml
Comment (chrislovecnm) [reply .bazelci/presubmit.yml:46]: I could generate a Fibonacci sequence or something.

File: examples/go-code-tutorial/BUILD.bazel
Comment (jibdugaw) [reply examples/go-code-tutorial/BUILD.bazel:1]: how about `examples/go-code` simply drop tutorial. Makes it clear and the tutorial cna rference the already extant example.

File: examples/go-code-tutorial/.gitignore
Comment (jibdugaw) [reply examples/go-code-tutorial/.gitignore:1]: recommend: add comment to the .gitignore file to make it clear why it exists. It's counter intuitive (per Alex's response) so pro-actively avoid a repeat question/confusion for future reviewers.

File: examples/go-code-tutorial/cmd/roll.go
Comment (jibdugaw) [reply examples/go-code-tutorial/cmd/roll.go:45]: Does providing a "finished" example make sense here? In the tutorial provide steps to build the Cobra CMD but in the source provide the finished output. Alternatively what about providing a tagged branch that applies the updated code to provide reasonable contents. I think giving a "working example" without muddying it with fill-in-the-blank comments makes it more straightforward. 

A thing to consider. Is the example going to be read directly for "how the frick do I do this" or the tutorial? On personal experience I tend to look at the source code directly unless I'm an absolutely rank-stank beginner.

**Review by jibdugaw — COMMENTED** [review]
See my earlier comment about directory names.

File: examples/go-code-tutorial/BUILD.bazel
Comment (chrislovecnm) [reply examples/go-code-tutorial/BUILD.bazel:1]: We are in rules_go, so I would assume it is go ... yah.  Do we have other ideas?

File: .bazelci/presubmit.yml
Comment (chrislovecnm) [reply .bazelci/presubmit.yml:46]: @alexeagle @fmeum any ideas on what package we should pull in? Also, should we fork the package, so that it does not disappear?

File: examples/go-code-tutorial/.gitignore
Comment (chrislovecnm) [reply examples/go-code-tutorial/.gitignore:1]: Will do!

File: examples/go-code-tutorial/cmd/roll.go
Comment (chrislovecnm) [reply examples/go-code-tutorial/cmd/roll.go:45]: I will remove the comments

File: .bazelci/presubmit.yml
Comment (chrislovecnm) [reply .bazelci/presubmit.yml]: I removed this and I am going to use klog

File: examples/go-code-tutorial/BUILD.bazel
Comment (chrislovecnm) [reply examples/go-code-tutorial/BUILD.bazel]: I changed the name to bazel-gazelle.

File: examples/go-code-tutorial/.gitignore
Comment (chrislovecnm) [reply examples/go-code-tutorial/.gitignore]: Done

**Review by alexeagle — APPROVED** [review]
LGTM

@fmeum want to squash & merge it?

**Review by fmeum — APPROVED** [review]
Looks good to me as well, thanks @chrislovecnm!

**Pull request merged**
